src/util/bson.hpp

```diff
@@ -0,0 +1,1218 @@
+    }
+
+    /// Whether or not this map is empty
+    bool empty()
+    {
+        return m_map.empty();
+    }
+
+    /// Pop the last entry of the map
+    void pop_back()
+    {
+        auto last_key = m_keys.back();
+        m_keys.pop_back();
+        m_map.erase(last_key);
+    }
+
+    std::vector<std::string> keys() const
+    {
+        return m_keys;
+    }
+private:
+    friend inline bool operator==(const IndexedMap<T>& lhs, const IndexedMap<T>& rhs);
+    std::unordered_map<std::string, T> m_map;
+    std::vector<std::string> m_keys;
+};
+
+template <typename T>
+inline bool operator==(const IndexedMap<T>& lhs, const IndexedMap<T>& rhs)
+{
+    return lhs.m_set == rhs.m_set && lhs.m_keys == rhs.m_keys;
+}
+
+/**
+ ====== BSON =======
+ */
+
+class Bson;
+
+/// A variant of the allowed Bson types.
+class Bson : public std::variant<
+    bson_null_t,
+    int32_t,
+    int64_t,
+    bool,
+    float,
+    double,
+    time_t,
+    const char*,
+    BinaryData,
+    Timestamp,
+    Decimal128,
+    ObjectId,
+    RegularExpression,
+    MinKey,
+    MaxKey,
+    IndexedMap<Bson>,
+    std::vector<Bson>
+> {
+public:
+    using base = std::variant<
+        bson_null_t,
+        int32_t,
+        int64_t,
+        bool,
+        float,
+        double,
+        time_t,
+        const char*,
+        BinaryData,
+        Timestamp,
+        Decimal128,
+        ObjectId,
+        RegularExpression,
+        MinKey,
+        MaxKey,
+        IndexedMap<Bson>,
+        std::vector<Bson>
+    >;
+    using base::base;
+    using base::operator=;
+};
+
+using BsonDocument = IndexedMap<Bson>;
+using BsonArray = std::vector<Bson>;
+
+
+template<class... Ts> struct overloaded : Ts... { using Ts::operator()...; };
+template<class... Ts> overloaded(Ts...) -> overloaded<Ts...>;
+
+/**
+ ======= JSON Serialization and Streams =======
+ */
+namespace {
+static void to_json(const Bson& bson, std::stringstream& ss)
+{
+    std::visit(overloaded {
+        [&ss](auto) {
+            ss << "null";
+        },
+        [&ss](bson_null_t) {
+            ss << "null";
+        },
+        [&ss](int32_t bson) {
+            ss << "{" << "\"$numberInt\"" << ":" << '"' << bson << '"' << "}";
+        },
+        [&ss](double bson) {
+            ss << "{" << "\"$numberDouble\"" << ":" << '"';
+            if (isnan(bson)) {
+                ss << "NaN";
+            } else if (bson == std::numeric_limits<double>::infinity()) {
+                ss << "Infinity";
+            } else if (bson == (-1 * std::numeric_limits<double>::infinity())) {
+                ss << "-Infinity";
+            } else {
+                ss << bson;
+            }
+            ss << '"' << "}";
+        },
+        [&ss](int64_t bson) {
+            ss << "{" << "\"$numberLong\"" << ":" << '"' << bson << '"' << "}";
+        },
+        [&ss](Decimal128 bson) {
+            ss << "{" << "\"$numberDecimal\"" << ":" << '"';
+            if (bson.is_nan()) {
```

Comment:
Is Decimal128::to_string() insufficient? It should handle everything you are doing here if you don't mind "Inf" and "-Inf"

Reply:
I do not believe Inf and -Inf are valid extjson, hence the conditionals.